tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_tensor_load_op.cc

```diff
@@ -0,0 +1,101 @@
+limitations under the License.
+==============================================================================*/
+
+// This file implements logic for lowering memref.tensor_load ops that are
+// inserted during `mhlo-legalize-to-lmhlo`.
+
+#include "mlir-hlo/Dialect/mhlo/transforms/passes.h"
+#include "mlir/Dialect/Shape/IR/Shape.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"  // TF:llvm-project
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassRegistry.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir {
+namespace lmhlo {
+namespace {
+
+using shape::ShapeOfOp;
+using tensor::ExtractOp;
+
+struct LegalizeTensorLoadOpPass
+    : public mlir::PassWrapper<LegalizeTensorLoadOpPass, FunctionPass> {
+  void getDependentDialects(DialectRegistry& registry) const override {
+    registry.insert<shape::ShapeDialect, tensor::TensorDialect,
+                    memref::MemRefDialect>();
+  }
+  // Perform the lowering to remove memref.tensor_load ops inserted during
+  // `mhlo-legalize-to-lmhlo`.
+  void runOnFunction() override;
+};
+
+// Converting:
+//   memref (operand) -> memref.tensor_load -> shape.shape_of
+//     to
+//   memref (operand) -> shape.shape_of
+struct ForwardShapeOfOp : public OpRewritePattern<ShapeOfOp> {
+  using OpRewritePattern<ShapeOfOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(ShapeOfOp shape_of,
+                                PatternRewriter& rewriter) const override {
+    auto tensor_load = shape_of.arg().getDefiningOp<memref::TensorLoadOp>();
+    if (!tensor_load) return failure();
+
+    rewriter.replaceOpWithNewOp<ShapeOfOp>(shape_of, shape_of.getType(),
+                                           tensor_load.memref());
+    return success();
+  }
+};
+
+// Converting:
+//   memref (operand) -> memref.tensor_load -> tensor.extract
+//     to
+//   memref (operand) -> memref.load
+struct ForwardExtractOp : public OpRewritePattern<ExtractOp> {
+  using OpRewritePattern<ExtractOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(ExtractOp extract,
+                                PatternRewriter& rewriter) const override {
+    auto tensor_load = extract.tensor().getDefiningOp<memref::TensorLoadOp>();
+    if (!tensor_load) return failure();
+
+    rewriter.replaceOpWithNewOp<memref::LoadOp>(
+        extract, extract.getType(), tensor_load.memref(), extract.indices());
+    return success();
+  }
+};
+
+void LegalizeTensorLoadOpPass::runOnFunction() {
+  auto func = getFunction();
+  auto context = &getContext();
+  OwningRewritePatternList patterns(context);
+  patterns.insert<ForwardShapeOfOp, ForwardExtractOp>(context);
+  if (failed(applyPatternsAndFoldGreedily(func, std::move(patterns))))
+    return signalPassFailure();
```

Comment:
Maybe drop the `return ` since `signalPassFailure` returns `void`?

Reply:
Done!